invisible_cities/cities/beersheba.py

```diff
@@ -0,0 +1,355 @@
+    cut_type        : Cut mode to the deconvolution output ('abs' or 'rel') using e_cut
+                      'abs': cut on the absolute value of the hits.
+                      'rel': cut on the relative value (to the max) of the hits.
+    inter_method    : Interpolation method.
+
+    Returns
+    ----------
+    apply_deconvolution : Function that takes hits and returns the
+    deconvolved data.
+    """
+    dimensions    = np.array(['X', 'Y', 'Z'][:n_dim])
+    sample_width  = np.array(sample_width   [:n_dim])
+    bin_size      = np.array(bin_size       [:n_dim])
+    diffusion     = np.array(diffusion      [:n_dim])
+
+    psfs          = pd.read_hdf(psf_fname)
+    deconvolution = deconvolve(n_iterations, iteration_tol, sample_width, bin_size, InterpolationMethod(inter_method))
+
+    e_type        = HitEnergy(energy_type)
+    cut_mode      = CutType(cut_type)
+    deco_mode     = DeconvolutionMode(deconv_mode)
+
+    def deconvolve_hits(df, z):
+        if   deco_mode is DeconvolutionMode.joint:
+            psf = psfs.loc[(psfs.z == find_nearest(psfs.z,                 z)) &
+                           (psfs.x == find_nearest(psfs.x, df.Xpeak.unique())) &
+                           (psfs.y == find_nearest(psfs.y, df.Ypeak.unique()))  , :]
+            deconv_image, pos = deconvolution(tuple(df.loc[:, v].values for v in dimensions), df.Q.values, psf)
+        elif deco_mode is DeconvolutionMode.separate:
+            psf_z0 = psfs.loc[(psfs.z == find_nearest(psfs.z,                 0)) &
+                              (psfs.x == find_nearest(psfs.x, df.Xpeak.unique())) &
+                              (psfs.y == find_nearest(psfs.y, df.Ypeak.unique()))  , :]
+            deconv_image, pos = deconvolution(tuple(df.loc[:, v].values for v in dimensions), df.Q.values, psf_z0)
+
+            dist     = multivariate_normal(np.zeros(n_dim), diffusion**2 * z / 10)
+            cols     = tuple(f"{v.lower()}r" for v in dimensions)
+            psf_cols = psf_z0.loc[:, cols]
+            gaus     = dist.pdf(psf_cols.values)
+            psf      = gaus.reshape(psf_cols.nunique())
+
+            deconv_image = nan_to_num(richardson_lucy(deconv_image, psf, n_iterations_g, iteration_tol))
+
+        return create_deconvolution_df(df, deconv_image.flatten(), pos, cut_mode, e_cut, n_dim)
+
+    def apply_deconvolution(df):
+        deco_dst = []
+        for peak, hits in df.groupby("npeak"):
+            hits.loc[:, "Q"] = hits.Q/hits.Q.sum()
+            if n_dim == 3:
+                deconvolved_hits =            deconvolve_hits(hits, weighted_mean_and_std(hits.Z, hits.E)[0])
+            else :
+                deconvolved_hits = pd.concat([deconvolve_hits(hits.loc[hits.Z == z, :], z) for z in hits.Z.unique()], ignore_index=True)
+
+            distribute_energy(deconvolved_hits, hits, e_type)
+            deco_dst.append(deconvolved_hits)
+
+        return pd.concat(deco_dst, ignore_index=True)
```

Comment:
After rewriting the function, the `apply` approach takes almost twice the time running (checked with cProfile) than the manual groupby-split, therefore I'm reluctant to do the `apply` approach.

Reply:
I recommend that you commit this approach, then follow it with a commit that reverts it. In the commit message titles provide clear indication that this was tried and rejected for performance reasons. In the body of the messages provide the profiling / benchmarking results. Keep it in the branch history after merging, even though the code doesn't go into production.

You want the log to document that this was tried, what *exactly* was tried, in a retrievable way. Like that:

1. The fact that it was tried and rejected is documented.
2. If we ever suspect the implementation that was tried, or don't trust the timings, it will be easy to reproduce.

Reply:
To add a bit more info on this, apply, at least on the current pandas version on IC, is called twice over the first group (https://stackoverflow.com/questions/21390035/pandas-groupby-apply-method-duplicates-first-group). This I imagine results in calling the deconvolution process an additional time and adding a humongous processing time.

However, newer versions of pandas do not longer show this behaviour (https://github.com/pandas-dev/pandas/pull/24748). I tested the code using this pandas version and the running time goes back to 'normal'. Therefore, either we update pandas (which could be delicate as I'm unsure if it would have any collateral on other functions) or we use the initial manual approach for splitting.

Reply:
Oh, I did not see your answer, @jacg. Will do as you propose!

Reply:
> either we update pandas (which could be delicate as I'm unsure if it would have any collateral on other functions) 

We need to keep up to date with the packages. Otherwise, everything will get more and more complicated to update in the future. No need to do it *right now*, but maybe we should implement a regular protocol to review packages and update them once a year or so. We should start by opening an issue and decide which packages to update.


As for the code, I am a bit annoyed that we cannot use `pandas`' built-ins, but maybe it's not so important...